File: ocaml/01-10/06.ml
(* Find out whether a list is a palindrome. (easy) *)

(* rev : 'a list -> 'a list *)
let rec rev = function
  | [] -> []
  | x::xs -> (rev xs) @ [x]

(* is_palindrome : 'a list -> bool *)
let is_palindrome xs = rev xs = xs

let _ = assert (is_palindrome [ "x" ; "a" ; "m" ; "a" ; "x" ])
let _ = assert ( not (is_palindrome [ "a" ; "b" ]))

File: prolog/01-10/06.pl
/*
P06 (*) Find out whether a list is a palindrome.
    A palindrome can be read forward or backward; e.g. [x,a,m,a,x].
*/
revacc([], Ys, Ys).
revacc([H|T], Ys, Acc) :- revacc(T, Ys, [H|Acc]).

rev([], []).
rev(Xs, Ys) :- revacc(Xs, Ys, []). 

palindrome([]).
palindrome(Xs) :- rev(Xs, Rs), Rs = Xs.
